pyquil/experimental/_program.py

```diff
@@ -0,0 +1,14 @@
+from typing import Optional
+from pyquil.quil import InstructionDesignator, Program
+
+
+class ExperimentalProgram(Program):
```

Comment:
Could you not update `Program` itself? Going from non-optional to optional should be backwards compatible

Reply:
Though I guess could result in a lot of refactoring to satisfy mypy

Reply:
Part of my intent in this PR is to not have to change _anything_ in the main library for simplicity, with this being almost a "draft" for the next major version (after a long period of iteration & feedback)

Reply:
That makes good sense 👍